benchmarking/smartseq2/confounding_factor_analysis/Confounding_factor_analysis.R

```diff
@@ -0,0 +1,325 @@
+# R script to do metrics test
+# loading library
+library(ggplot2)
+library(MASS)
+library(plyr)
+library(reshape2)
+library(ggpubr)
+library(gridExtra)
+library(ggpmisc)
+library(cowplot)
+library(corrplot)
+library(ggrepel)
+library(optparse)
+library(rsvd)
+set.seed(42)
+
+# if p<0.05, return 1 value.
+convert2SigLevels <- function(P) {
+  x <- floor(-1 * log(P, base = 10))
+  x[x > 1] <- 1
+  return(x)
+}
+# cover to foldchnages
+# mat1 and mat2 are two data matrix
+foldchanges <- function(mat1, mat2) {
+  # log2 transformation
+  # first 2 columns are gene ID and length
+  logmd1 <- log(mat1[, -c(1:2)] + 1, base = 2)
+  # match gene ID
+  mlist <- match(mat1[, 1], mat2[, 1])
+  # match sample column
+  nlist <- match(colnames(mat1), colnames(mat2))
+  mat2 <- mat2[mlist, nlist]
+  # log2 transformation
+  logmd2 <- log(mat2[mlist, -c(1:2)] + 1, base = 2)
+  fcmat <- logmd1 - logmd2
+  out <- cbind(mat1[, 1:2], fcmat)
+  return(out)
+}
+# log2 transformation
+takelog2 <- function(mat) {
+  dd <- log(mat[, -c(1:2)] + 1, base = 2)
+  out <- cbind(mat[, c(1:2)], dd)
+  return(out)
+}
+# correlation between qc metrics and data matrics.
+# first run PCA on data matrix and top top N PCs
+# Then calculate correlation and correlation test between
+# QC metrics and N PCs
+# Visualize this correlation and test results in corrplot
+# Y-axis rank by total number of significant correlation test
+# QCmetrics with High rank(on the top) indicates these metrics have
+# high impact on data matrix
+# input cmd: wither rpca or pca. rpca is randomized PCA
+CorrQCvsPCs <- function(qc_mets, cnts, npcs, output_name, cmd) {
+  if (cmd == "rpca") {
+    cnt.pca <- rpca(cnts[, -c(1:2)], scale = T)
+  } else{
+    cnt.pca <- prcomp(cnts[, -c(1:2)], scale = T)
+  }
+  # take top npcs from rotation matrix
+  pcs <- cnt.pca$rotation[, 1:npcs]
+  nc1 <- ncol(qc_mets)
+  # merge qc metrics with pcs by rowname which is sample ID
+  dt <- merge(qc_mets, pcs, by = 0)
+  # run correlation test
+  res <- cor.mtest(dt[, -1])
+  pmat <- res$p # extract p values
+  # pmat is square matrix
```

Comment:
How do you know this is a square matrix if the number of PCs is an input that doesn't match the number  of qc metrics? Am I misunderstanding something here?

Reply:
cor.mtest is self-correlation test which will return N*N matrix, N is sum of number of QC metrics + number of PCs. then I only select correlation between QC to PCs, which is subset of this matrix.

Reply:
`cor.mtest(x)` will return square matrix N*N, N is sum of # of QCs +#  of PCs. later, I only select the subset of this matrix.